### PR TITLE
Warn on failure to end warmup, add explicit api for start of model invocation

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101129

CUDAGraph trees needs to known when you are doing a new invocation of your model. We have two heuristics for that : 
- you invoke torch.compile again (like as a top level module you compiled)
- you have run a forward with a corresponding backward that hasn't been invoked yet, in which case we ignore the above

This doesn't always get it right, especially if you forget to use torch.no_grad() in inference. This adds a warning for that case, and adds an explicit `cudagraph_mark_step_begin` api.

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

File: torch/_inductor/cudagraph_trees.py [hunk 1]
```diff
@@ -1869,6 +1916,40 @@ def try_end_curr_warmup(self):
...
+
+        # first node executed or same pattern
+        if prior_warmup.parent is None or (
+            prior_warmup.parent.wrapped_function == self.current_node.wrapped_function
+        ):
+            self.warned_functions.add(function_id)
+            warnings.warn(
+                "Unable to end warm up phase of CUDAGraphs because of pending, uninvoked backwards. "
+                "Consider running with torch.no_grad() or using torch._inductor.cudagraph_mark_step_begin() "
+                "before each model invocation"
```
Comment (ezyang) [inline torch/_inductor/cudagraph_trees.py:1950]: I feel like this can be worded more clearly

Comment (ezyang) [reply torch/_inductor/cudagraph_trees.py:1950]: Imagine the user hasn't read the manual. What is "warm up phase"? What exactly is the user supposed to do? This might be a good place to link to a wiki page which we can more easily evolve over time.

Comment (eellison) [reply torch/_inductor/cudagraph_trees.py]: Rephrased, will link to doc when it lands
